### DISCUSSION TITLE
AC Typed Schema files invalid against JSON Schemas (and more errors)

### DISCUSSION THREAD

**Discussion by jamesthurley** [microsoft/AdaptiveCards#8943] [General]
For example this file:

https://github.com/microsoft/AdaptiveCards/blob/main/schemas/src/enums/ImageSize.json

Says it uses this schema:

https://raw.githubusercontent.com/microsoft/AdaptiveCards/6f39aedce45864ae1067ed44a5551dc973790bb5/source/nodejs/typed-schema/schema/lib/Type.json

But it isn't valid against that schema. It contains an extra `description` property, and is missing the `type` property. Have the schemas for the `ac-typed-schema` JSON been abandoned? Would a PR to update them be appreciated?

**Comment by jamesthurley** [comment]
Following up, I've noticed some of the sample AdaptiveCards are also not valid, for example:

https://github.com/microsoft/AdaptiveCards/blob/main/samples/v1.0/Elements/Container.Style.json

This one contains a `style` field on the root AdaptiveCard, which is not valid by the schema it references.

Another example is:

https://github.com/microsoft/AdaptiveCards/blob/main/samples/v1.1/Tests/Image.Stretch.json

Which has a `height` field on the `Column` elements, which is not valid.

**Comment by jamesthurley** [comment]
I ran your own tests in `source/nodejs/tests/test-adaptive-card-schema/src/test-adaptive-card-schema.ts` in this repository and they also report these sample/test cards as being invalid, along with plenty others.

```
../../../../samples/v1.0/Elements/Action.ShowCard.Style.json: INVALID
../../../../samples/v1.0/Elements/Action.ShowCard.json: Success
...
../../../../samples/v1.0/Elements/Container.Spacing.json: Success
../../../../samples/v1.0/Elements/Container.Style.json: INVALID
../../../../samples/v1.0/Elements/Container.json: Success
../../../../samples/v1.0/Elements/Fact.json: Success
...
../../../../samples/v1.0/Tests/AdaptiveCard.UnknownElements.json: Success (correctly invalid)
../../../../samples/v1.0/Tests/AdditionalProperty.json: Success (correctly invalid)
...
../../../../samples/v1.1/Tests/Image.ResouceResolver.json: Success
../../../../samples/v1.1/Tests/Image.Stretch.json: INVALID
../../../../samples/v1.1/Tests/ImageSet.Stretch.json: INVALID
...
```

Some are reported as "correctly invalid" as the test is asserting that these are invalid as expected, but the ones marked as INVALID are actual failing tests. I guess these tests just aren't run any more.

Unfortunately it doesn't give a great impression to someone new coming to the repo to find the tests are broken, the schemas are out of date, the samples are incorrect.

I was planning on replicating your test suite to validate my parser, using your `shouldFail` list in `test-adaptive-card-schema.ts` to check that cards which should fail do indeed fail. I've now reverted to validating each sample card against the JSON Schema and using that to determine which ones should fail instead, which eliminates more of the samples. This seems like a good workaround for now.

**Comment by jamesthurley** [comment]
Ok, relying on validating the samples against the AdaptiveCards schema to determine if I should be able to parse them doesn't work either, due to errors in the current live schema.

In the current schema here: https://adaptivecards.io/schemas/adaptive-card.json

Both `Container` and `TableCell` have the same error. The `rtl` property is defined as:

```
         "rtl?": {
          "anyOf": [
            {
              "type": "boolean"
            },
            {
              "type": "null"
            }
          ],
          "version": "1.5"
        },
```

This property should be `rtl` not `rtl?`. This error means samples such as `v1.5/Tests/RtlTestCard.json` currently fail validation when they should pass.

Meanwhile `v1.5/Tests/Container.ScrollableSelectableList.json` contains duplicate "type" and "version" fields on the root object (they are at the top and repeated at the bottom down on lines 568 and 569. This makes them technically valid but Rust's Serde complains of duplicate fields.

**Comment by jamesthurley** [comment]
In the current live schema ChoiceInputStyle is defined as:
```
"ChoiceInputStyle": {
      "description": "Style hint for `Input.ChoiceSet`.",
      "anyOf": [
        {
          "enum": [
            "compact",
            "expanded"
          ]
        },
        {
          "pattern": "^([c|C][o|O][m|M][p|P][a|A][c|C][t|T])|([e|E][x|X][p|P][a|A][n|N][d|D][e|E][d|D])$"
        }
      ]
    }
```

But it should also allow "filtered" since version 1.5:

ChoiceInputStyle.json:
```
{
	"$schema": "https://raw.githubusercontent.com/microsoft/AdaptiveCards/6f39aedce45864ae1067ed44a5551dc973790bb5/source/nodejs/typed-schema/schema/lib/Type.json",
	"classType": "Enum",
	"description": "Style hint for `Input.ChoiceSet`.",
	"values": [
		"compact",
		"expanded",
		{
			"value": "filtered",
			"description": "Allows users to filter choices in a choice set.",
			"version": "1.5"
		}
	]
}
```

This causes `v1.5/Tests/Input.ChoiceSet.FilteredStyle.TestCard.json` to incorrectly fail validation.

**Comment by jamesthurley** [comment]
The final thing that tripped me up was that `v1.5/Tests/Image.Svg.json` fails validation because the final url...

```
data:image/svg+xml;utf8,<svg width=\"120\" height=\"120\" version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" style=\"fill: none; stroke: black\"> <path d=\"M 10,10 l 100,0 l 0,100 l -100,0 l 0,-100 l 100,100 M 110,10 l -100,100\" /> </svg>
```

... doesn't seem to validate as a `uri-reference` either with the Rust `jsonschema` crate or with JSON.NET Schema.

However, after fixing the above issues locally I seem to be able to parse all the tests and samples which validate against your (fixed) schema, and I fail to parse all the ones that fail validation 🎉

**Comment by jamesthurley** [comment]
I tried to run `cargo typify` on the Host Config schema located in `schemas/host-config.json`, which revealed some errors in this schema as well.

With the following fixed I can successfully generate rust types for this schema.

# 1
```
definitions -> FactSetConfig -> properties -> value -> default -> weight
```
Current value is `default`, which is invalid. It should be `normal`.

# 2
```
definitions -> InputLabelConfig -> properties -> weight
```

The `default` field's value of `default` is not valid (lowercase first letter) according to the enum values of `Lighter`, `Default` and `Bolder`. For consistency with the rest of the schema I would change the enum values to the camel case `lighter`, `default` and `bolder`.

# 3
```
definitions -> ShowCardConfig -> properties -> style
```

Currently this is defined as:
```
{
	"$ref": "#/definitions/ContainerStyleConfig",
	"default": "emphasis"
}
```

But judging by the samples it should actually be:
```
{
	"type": "string",
	"default": "emphasis",
	"enum": [
		"default",
		"emphasis"
	]
}
```

**Comment by jamesthurley** [comment]
The `host-config.schema` is also missing the following at the root:

```
"$ref": "#/definitions/HostConfig"
```

Which is required to make it validate a host config. With that in many (all) of the host config samples fail validation because the casing is incorrect on many enum values.

After fixing the casing the sample will pass, but only because the schema doesn't specify for any definition:

```
"additionalProperties": false
```

Once you start adding those in you get more failures, as it looks like the host config schema hasn't been kept up to date with the samples or what is allowed in a host config.

**Comment by jamesthurley** [comment]
A different type of error not caught due to missing `"additionProperties": false`.

In the `microsoft-teams-light.json` sample, in `factSet.title` and `factSet.value`, both specify `"warp": true`. But `warp` is not a valid property, it should be `wrap`.

This is also an error in `microsoft-teams-dark.json` and `cortana-skills-dark.json`.

**Comment by cwacek** [comment]
Just to chime in here, I think the AdaptiveCard [schema](https://adaptivecards.io/schemas/1.5.0/adaptive-card.json) is invalid as well. The `Action.Execute` object defines a whole bunch of properties with no specification including `title`, `requires`. [Here](https://github.com/cwacek/python-jsonschema-objects/blob/b3316cf4e1ba3327b780f2942c8edaf4a7d54b3e/test/resources/adaptive-card.json#L35-L43) is a direct link to the offending definitions all of which prevent that library from loading the schema. 

@jamesthurley has anyone acknowledged this is an issue?

**Comment by jamesthurley** [comment]
@cwacek No, not heard anything yet. Perhaps these discussions aren't monitored by anyone. My current plan is to loop back in 6 months, when I'll have a bit more spare time, and submit some PRs. I'd do it sooner but the repo looks quite complicated with a bunch of code generation and tooling in various languages, and I want to take the time to make sure I understand it all.

In the mean time I'm just working off fixed up local copies of the files I need.
